### ISSUE TITLE
[clang] False positive: -Wmissing-noreturn triggered on lambdas

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#154493]
Example:

```cpp
#include <stdexcept>

int main()
{
    auto l = []() {
        throw std::runtime_error{"Error."};
    };
}
```

https://godbolt.org/z/zdWdfMdxM

Expectation is to get no warning, since as a user I cannot add `[[noreturn]]` to the internal `operator()` of the lambda implementation.

**Comment by carlosgalvezp** [comment]
In that case the warning should only fire in C++23 onwards. The user cannot act upon this warning earlier than that.

**Comment by carlosgalvezp** [comment]
The following change to `SemaDeclAttr.cpp` fixes the issue, if you think it's acceptable I can put up a PR!

```cpp
+   // [[noreturn]] can only be added to lambdas since C++23
+   if (const auto *MD = dyn_cast<CXXMethodDecl>(FD))
+    {
+      if (!S.getLangOpts().CPlusPlus23 && isLambdaCallOperator(MD))
+        return;
+    }

    // Emit a diagnostic suggesting the function being marked [[noreturn]].
    S.Diag(FD->getLocation(), diag::warn_suggest_noreturn_function)
        << /*isFunction=*/0 << FD;
```

**Comment by AaronBallman** [comment]
> The following change to `SemaDeclAttr.cpp` fixes the issue, if you think it's acceptable I can put up a PR!
> ```
> +   // [[noreturn]] can only be added to lambdas since C++23
> +   if (const auto *MD = dyn_cast<CXXMethodDecl>(FD))
> +    {
> +      if (!S.getLangOpts().CPlusPlus23 && isLambdaCallOperator(MD))
> +        return;
> +    }
> 
>     // Emit a diagnostic suggesting the function being marked [[noreturn]].
>     S.Diag(FD->getLocation(), diag::warn_suggest_noreturn_function)
>         << /*isFunction=*/0 << FD;
>```

I think that's a reasonable approach, though I would do:
```
+   // [[noreturn]] can only be added to lambdas since C++23
+   if (const auto *MD = dyn_cast<CXXMethodDecl>(FD);
+       MD && !S.getLangOpts().CPlusPlus23 && isLambdaCallOperator(MD))
+      return;
```
